text/src/autogluon/text/automm/predictor.py

```diff
@@ -569,21 +569,83 @@ def _fit(
 
         if trainer.global_rank == 0:
             top_k_avg_ckpt_path = os.path.join(save_path, "model.ckpt")
+            # Obtain the best K models sorted from the best performance to the worst performance
+            best_k_models_sorted = sorted(list(checkpoint_callback.best_k_models.items()),
+                                          key=lambda ele: ele[1], reverse=(minmax_mode == 'max'))
+            best_k_models_path = [ele[0] for ele in best_k_models_sorted]
+
+            # TODO(?) Currently, all top-K checkpoints are loaded to the main memory.
+            #  If the checkpoints are really large (e.g., super-giant backbones), this operation can take a lot of
+            #  CPU memory and may potentially trigger memory issue. Consider to optimize that using the
+            #  offload strategy.
             all_state_dicts, ckpt_template = gather_top_k_ckpts(
                 ckpt_dir=save_path,
-                ckpt_paths=checkpoint_callback.best_k_models.keys(),
+                ckpt_paths=best_k_models_path,
             )
+            if config.optimization.top_k_average_method == 'union_soup':
+                logger.info(f'Start to fuse {config.optimization.top_k} checkpoints via the UnionSoup algorithm.')
+                ingredients = all_state_dicts
+            else:
+                if not best_k_models_path:
+                    # In the case of ddp_spawn, the best_k_models_path can be empty. This is due the
+                    # limitation of PT Lightning: https://github.com/PyTorchLightning/pytorch-lightning/issues/5582
+                    # Thus, we will need to reevaluate the validation performance of these checkpoints
+                    candidate_performance = []
+                    for cand_state_dict in all_state_dicts:
+                        self._model = self._load_state_dict(
+                            model=model,
+                            state_dict=cand_state_dict,
+                        )
+                        candidate_performance.append(self.evaluate(val_df,
+                                                                   [validation_metric_name])[validation_metric_name])
+                    all_state_dicts = [v[0] for v in sorted(zip(all_state_dicts, candidate_performance),
+                                                            key=lambda ele: ele[1], reverse=(minmax_mode == 'max'))]
+                if config.optimization.top_k_average_method == 'greedy_soup':
+                    # Select the ingredients based on the methods proposed in paper
+                    #  "Model soups: averaging weights of multiple fine-tuned models improves accuracy without
+                    #  increasing inference time", https://arxiv.org/pdf/2203.05482.pdf
+                    monitor_op = {"min": operator.lt, "max": operator.gt}[minmax_mode]
```

Comment:
Should they be `operator.le` and `operator.ge`?

Reply:
Done.